p11kit/attribute.go

```diff
@@ -135,7 +135,36 @@ func signECDSA(priv crypto.Signer, m mechanism, data []byte) ([]byte, error) {
 	if !m.noParams() {
 		return nil, fmt.Errorf("CKM_ECDSA does not take any parameters: %w", errArgumentsBad)
 	}
-	return priv.Sign(rand.Reader, data, crypto.Hash(0))
+	sig, err := priv.Sign(rand.Reader, data, crypto.Hash(0))
+	if err != nil {
+		return nil, err
+	}
+
+	// PKCS11 use raw R, S value as signature, priv.Sign return a DER encoded R, S sequence
+	// convert DER sequence to raw R, S
+	// http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cs01/pkcs11-curr-v2.40-cs01.html#_Toc399398879
+	var sigDecode []*big.Int
+	_, err = asn1.Unmarshal(sig, &sigDecode)
+	if err != nil {
+		return nil, err
+	}
+
+	// Padding, both big.Int and PKCS#11 signature are big-endian
+	r, s := sigDecode[0].Bytes(), sigDecode[1].Bytes()
+	if len(r) > len(s) {
+		size := len(r)
```

Comment:
Can we compute size from the public key?

```
pub, ok := priv.Public().(*ecdsa.PublicKey)
if !ok {
    return nil, fmt.Errorf("expected ecdsa public key, got %T", priv.Public())
}
size := pub.Params().N.BitLen()/8
```

Reply:
The signature is double size of private key. I'm not sure whether public key and private key size are same. If it is, we can use this approach.

Reply:
From the link in your comment:

> For the purposes of these mechanisms, an ECDSA signature is an octet string of even length which is at most two times nLen octets, where nLen is the length in octets of the base point order n.

So it should be determined by the curve (instead of D), which is shared with public key :)